nimbus/db/core_db/core_apps_newapi.nim

```
@@ -921,6 +921,21 @@ proc haveBlockAndState*(db: CoreDbRef, headerHash: Hash256): bool =
   # see if stateRoot exists
   db.exists(header.stateRoot)
 
+proc getBlockWitness*(db: CoreDbRef, blockHash: Hash256): seq[byte] {.gcsafe.} =
+  let data = db.newKvt(Shared)
+               .get(blockHashToBlockWitnessKey(blockHash).toOpenArray).valueOr:
+    if error.error != KvtNotFound:
+      warn logTxt "getBlockWitness()", blockHash, action="get()", error=($$error)
+    return
```

Would it not make more sense to return here a `Result` or `results.Opt`?

---

It appears that the db.newKvt.get proc returns the CoreDbRc type which is defined deeper inside the database layer code. It looks like none of the other procs in this file are returning this type perhaps to hide this type from the calling code. I copied this code pattern from the getScore proc defined further above. I agree that it doesn't look great to just return and appears to ignore errors apart from the warning. Perhaps it would be better let the exception be thrown up the stack here if one occurs.

---

Hi @mjfh , are you able to help with addressing the question above? I see you are the original author of this file so I'm wondering what the reasoning is for this code pattern?

---

> Perhaps it would be better let the exception be thrown up the stack here if one occurs?

I had a look at this code and it doesn't seem any exception (CatchableError) can occur here. The `db.newKvt.get` code does not raise at least (see `{.push raises: [].}` on top of that file). And the `CoreDbRc` returned is a Result type (You are already using it like that, with the `valueOr`).

So you could just return that same result type. But perhaps better would be to remap the current Result type to a simpler one without the `CoreDbErrorRef` error part and instead something more digestible for the API user (like just an error string). Unless it would be required for the layer up to have that error information, but I don't think so.

> I copied this code pattern from the getScore proc defined further above.

Yes, so `getScore` results in 0 basically when the getter fails. I don't know of this functionality, but perhaps there the idea is to get 0 if no score exists yet in the db (no idea really, someone familiar with this code could say if this is an issue or not).
In the case of `getBlockWitness` it would require to do a `len>0` test after the call or rely on failing at the moment of verification (but that could be after a network transfer for example).

---

Sure, I'll update it to use the regular result type and will map the error code to a plain string.